### PR TITLE
[SPARK-48]  Unit Tests for mechaclient.py

### PR DESCRIPTION
**Pull request by shatteredbeam**
Unit tests for mechaclient.py

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -315,4 +316,4 @@ def test_fact_fx() -> Fact:
                 editedby='Shatt',
                 mfd=False,
                 edited=None
-                )
+                )
```
Comment (theunkn0wn1) [inline tests/conftest.py]: missing \n at EOL

File: tests/test_mechaclient.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+pytestmark = pytest.mark.mechaclient
+
+
+def test_verify_version(bot_fx):
+    """
+    Asserts the version is set, and returning string.
+    """
+    assert bot_fx.__version__
+    assert isinstance(bot_fx.__version__, str)
```
Comment (theunkn0wn1) [inline tests/test_mechaclient.py]: [scope] testing builtin.
[error] testing behavior of hardcoded attribute.

File: tests/conftest.py [hunk 3]
```diff
@@ -37,6 +37,7 @@
 
 setup_logging("logs/unit_tests.log")
 
+from src.mechaclient import MechaClient
```
Comment (Eearslya) [inline tests/conftest.py]: [Style] (unused-import) What does adding this import provide? I can't see any additional code that uses this import.

File: tests/test_mechaclient.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+    Asserts the Fact Manager can be set, requires a FactManager object, and returns properly.
+    """
+    # Helper Class
+    class FakeFactManager(FactManager):
+        # Overriding the parent class init to prevent it from creating a DB connection.
+        def __init__(self):
+            ...
+
+    bot_fx.fact_manager = FakeFactManager()
+    assert isinstance(bot_fx._fact_manager, FakeFactManager)
```
Comment (Eearslya) [inline tests/test_mechaclient.py:78]: [Discussion] Should these tests be separated? Given the parametrize, it's testing both a good path and a bad path with each invocation.

File: tests/test_mechaclient.py [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+
+@pytest.mark.parametrize("rat_board_bad_types", ["Something that isn't a Rat Board", (1, 3), 3.14])
+def test_verify_rat_board(bot_fx, rat_board_bad_types):
+    """
+    Asserts the Rat Board can be set, requires a Rat Board object, and returns properly.
+    """
+    bot_fx.board = RatBoard()
+
+    assert isinstance(bot_fx._rat_board, RatBoard)
+    assert bot_fx.board is bot_fx._rat_board
```
Comment (Eearslya) [inline tests/test_mechaclient.py:115]: [Discussion] Should these tests be separated? Given the parametrize, it's testing both a good path and a bad path with each invocation.

File: pytest.ini [hunk 6]
```diff
@@ -23,6 +23,7 @@ markers =
     ratsignal_parse: marks a test as a ratsignal parser test
     autocorrect: tests for the autocorrect util
     offline: offline-aware tests
+    mechaclient: marks a test as a mechaclient test.
```
Comment (theunkn0wn1) [inline pytest.ini]:
```suggestion
    mechaclient: mechaclient tests
```

File: tests/conftest.py (hunk 3 above)
Comment (theunkn0wn1) [reply tests/conftest.py]: Id argue this is convention / refactor more than it is style, but yes, this appears to be a needless import.

File: tests/test_mechaclient.py [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+def test_verify_rat_board_private(bot_fx):
+    """
+    Asserts the private reference for the Rat Board is none, after instantiation.
+    """
+    assert bot_fx._rat_board is None
+
+
+def test_verify_api_handler(bot_fx):
+    """
+    Asserts the api_handler property returns an object, and cannot be set (no setter)
```
Comment (theunkn0wn1) [inline tests/test_mechaclient.py]: [error] this is not what the test is doing.

File: tests/test_mechaclient.py [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+    """
+    # Helper Class
+    class FakeAPIHandler:
+        ...
+
+    assert bot_fx.api_handler is None
```
Comment (theunkn0wn1) [inline tests/test_mechaclient.py:59]: [error] this is already tested in a different test case.

**Review by theunkn0wn1 — CHANGES_REQUESTED** [review]

File: tests/conftest.py (hunk 3 above)
Comment (shatteredbeam) [reply tests/conftest.py]: If its an invalid import, I'll remove it.  I'll double check.

File: tests/test_mechaclient.py (hunk 8 above)
Comment (shatteredbeam) [reply tests/test_mechaclient.py:59]: Not specifically this one, removing the test removes coverage for the api_handler getter.

File: tests/test_mechaclient.py (hunk 7 above)
Comment (shatteredbeam) [reply tests/test_mechaclient.py]: Docstring fixed in 96dac7b.

File: pytest.ini (hunk 6 above)
Comment (shatteredbeam) [reply pytest.ini]: Changed to 'mechaclient tests' in 96dac7b.

File: tests/test_mechaclient.py (hunk 2 above)
Comment (shatteredbeam) [reply tests/test_mechaclient.py]: Superfluous test removed in 96dac7b.

File: tests/conftest.py (hunk 3 above)
Comment (shatteredbeam) [reply tests/conftest.py]: unused import removed in 96dac7b.

File: tests/test_mechaclient.py (hunk 5 above)
Comment (shatteredbeam) [reply tests/test_mechaclient.py:115]: They could be separated, but what would be gained from splitting it?  The tests do pass with known good or bad data.

Comment (Eearslya) [reply tests/test_mechaclient.py:115]: It's more of a convention thing IMO, making it so one test tests one thing. No real tangible benefit other than organization. (Which is still nice to have.)

File: tests/conftest.py (hunk 3 above)
Comment (Eearslya) [reply tests/conftest.py]: I believe you removed the wrong import; the one I was referring to is in `conftest.py`, which is still there.

File: tests/test_mechaclient.py (hunk 5 above)
Comment (theunkn0wn1) [reply tests/test_mechaclient.py:115]: Arguably, tests should only test one scenario

Either test the good path, or the bad path, but not both within a test definition.

File: tests/conftest.py (hunk 3 above)
Comment (theunkn0wn1) [reply tests/conftest.py]: This resolved?

**Review by theunkn0wn1 — APPROVED** [review]
LGTM

**Review by Eearslya — APPROVED** [review]

**Pull request merged**